x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+// distributeRewards is the main function for reward distribution for providers and delegators
+func (k Keeper) distributeRewards(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, block uint64, totalReward math.Int) error {
+	epoch, _, err := k.epochStorageKeeper.GetEpochStartForBlock(ctx, block)
+	if err != nil {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", err,
+			utils.Attribute{Key: "block", Value: block},
+		)
+	}
+	stakeStorage, found := k.epochStorageKeeper.GetStakeStorageEpoch(ctx, epoch, chainID)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("stake storage for epoch not found"),
+			utils.Attribute{Key: "epoch", Value: epoch},
+			utils.Attribute{Key: "chainID", Value: chainID},
+		)
+	}
+
+	stakeEntry, found, stakeIndex := k.epochStorageKeeper.GetStakeEntryByAddressFromStorage(ctx, stakeStorage, providerAddr)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, stakeEntry, stakeIndex, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot calculate provider reward and update delegator reward map", err)
+	}
+
+	providerRewardCoins := sdk.Coins{sdk.NewCoin(epochstoragetypes.TokenDenom, providerReward)}
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, providerAddr, providerRewardCoins)
+	if err != nil {
+		// panic:ok: reward transfer should never fail
+		utils.LavaFormatPanic("critical: failed to send reward to provider", err,
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "reward", Value: providerRewardCoins},
+		)
+	}
+
+	delegations, err := k.dualStakingKeeper.GetProviderDelegators(ctx, providerAddr.String(), epoch)
+	if err != nil {
+		return utils.LavaFormatError("cannot get provider's delegators", err)
+	}
+
+	var relevantDelegations []dualstakingtypes.Delegation
```

Comment:
you can use `utils/slices`'s Filter() helper.

Reply:
It's actually a `Map()` function, not filter. 
See #785

Reply:
merged main and solved conflicts. Now using your implementation. see https://github.com/lavanet/lava/commit/8261db8d95f5cc75be14d2d0ff3bda802b5f7eee